src/hd.rs

```diff
@@ -181,111 +181,59 @@ impl HDWallet {
-                network,
-                xpub.script_type,
-                gap_limit,
-                initial_import_size,
-                rescan_policy,
-            ),
-            // internal chain (change)
-            Self::new(
-                xpub.extended_pubkey
-                    .derive_pub(&*EC, &[ChildNumber::from(1)])?,
-                network,
-                xpub.script_type,
-                gap_limit,
-                initial_import_size,
-                rescan_policy,
-            ),
-        ])
-    }
-
-    pub fn from_xpubs(
-        xpubs: &[(XyzPubKey, RescanSince)],
-        bare_xpubs: &[(XyzPubKey, RescanSince)],
+    pub fn from_descriptors(
+        descriptors: &[(String, RescanSince)],
         network: Network,
         gap_limit: u32,
         initial_import_size: u32,
+        rpc: &RpcClient,
     ) -> Result<Vec<Self>> {
         let mut wallets = vec![];
-        for (xpub, rescan) in xpubs {
-            wallets.append(
-                &mut Self::from_xpub(
-                    xpub.clone(),
-                    network,
-                    gap_limit,
-                    initial_import_size,
-                    *rescan,
-                )
-                .with_context(|| format!("invalid xpub {}", xpub))?,
-            );
-        }
-        for (xpub, rescan) in bare_xpubs {
+        for (descriptor, rescan) in descriptors {
             wallets.push(
-                Self::from_bare_xpub(
-                    xpub.clone(),
+                Self::from_descriptor(
+                    descriptor.clone(),
                     network,
                     gap_limit,
                     initial_import_size,
                     *rescan,
+                    rpc,
                 )
-                .with_context(|| format!("invalid xpub {}", xpub))?,
+                .with_context(|| format!("invalid descriptor {}", descriptor))?
+                .clone(),
             );
         }
         if wallets.is_empty() {
-            warn!("Please provide at least one xpub to track (via --xpub or --bare-xpub).");
-            bail!("no xpubs provided");
+            warn!("Please provide at least one descriptor to track (via --descriptors).");
+            bail!("no descriptors provided");
         }
         Ok(wallets)
     }
 
-    pub fn derive(&self, index: u32) -> ExtendedPubKey {
-        self.master
-            .derive_pub(&*EC, &[ChildNumber::from(index)])
-            .unwrap()
+    pub fn from_descriptor(
+        descriptor: String,
+        network: Network,
+        gap_limit: u32,
+        initial_import_size: u32,
+        rescan_policy: RescanSince,
+        rpc: &RpcClient,
+    ) -> Result<Self> {
+        let descriptor = Descriptor::new(&descriptor, rpc).unwrap();
+        // FIXME
+        //ensure!(
+        //xpub.matches_network(network),
+        //"xpub network mismatch, {} is {} and not {}",
+        //xpub,
+        //xpub.network,
+        //network
+        //);
```

Comment:
Not easy to dig inside the descriptor and make assertions about the pubkeys ...

Reply:
I think its fine if we don't, bitcoind will verify this for us.